tests/checkpointing/test_model_checkpoint.py

```diff
@@ -1202,3 +1202,32 @@ def test_check_val_every_n_epochs_top_k_integration(tmpdir):
     )
     trainer.fit(model)
     assert set(os.listdir(tmpdir)) == {"epoch=1.ckpt", "epoch=3.ckpt"}
+
+
+def test_model_checkpoint_attributes(tmpdir):
```

Comment:
I think it's better to update this existing test with all the arguments that are being restored.
https://github.com/PyTorchLightning/pytorch-lightning/blob/5576fbc5f9a7d0bc71ad26b8b54775110e675808/tests/models/test_restore.py#L269-L272

Reply:
Imo we could differentiate between a:
* Unit tests that checks whether states are written to disk properly (i.e. testing `ModelCheckpoint.on_save_checkpoint` and `ModelCheckpoint.on_load_checkpoint`).
* The existing functional test that checks whether model states are restored properly when providing `ckpt` to `trainer.fit`.

Regarding the attbributes that are restored: Why don't we restore `current_score`, `monitor` and `dirpath` attributes that are written to `ckpt`. Is there an explicit reason for this ?

Reply:
`current_score` isn't required to be reloaded because it will eventually change right after the next call, and for `dirpath`, even while resuming it can be different and there won't be any problem. For `monitor`, I guess it can be changed too since we don't force users to use the old one but then best_model_path/score would be irrelevant.

Regarding unit testing, I am just concerned about repetitive tests. We test that while resuming `on_save/load_checkpoint` is called, elsewhere, all we need to test what's being saved and reloaded which I guess the existing test covers up.

Reply:
I definitely prefer unit testing here. There are already plenty of tests that ensure that "checkpointing" works. 
Here, we would only want to instantiate the ModelCheckpoint and call the hooks to test their function (no trainer.fit, no loading checkpoint).

Reply:
I'd been fine with adding unit tests here only if there wasn't any existing test here that checks what we need to but adding another similar test makes it hard later to track which tests to update in case some new stuff gets added just like we missed here and later found out there's already a similar existing test.

Reply:
Hey, I just fixed the broken funciontality test `test_callbacks_state_fit_ckpt_path` (`ModelCheckpoint` actually never saved/triggered, since trainer was fitted with `max_steps = 1`  and `ModelCheckpoint` configured to trigger after every epoch).

However, tests fail now. I inspected states manually and `before` states are updated properly now, whereas `after` states remain empty/uninitialized (before my fix, both `before` and `after` states were same as upon their initialization). Any ideas why `before` isn't restored properly for the test ? This might hint at another bug somewhere in the code.

I would afterwards extend the test with the additional keys and then refactor the unit-test to do pure testing of `ModelCheckpoint.on_save_checkpoint` and `ModelCheckpoint.on_load_checkpoint`

Reply:
Hey @ORippler ,

You are correct and the test did not do what it was supposed to do any more. Updating `max_steps` to `max_epochs` seems correct to me (maybe we could additionally limit the number of batches here to 1 or 2 to make it faster?)

After adjusting the test it failed, because `save_on_train_epoch_end` was not specified and thereby internally set to True. However this caused the restore to fail since the state_keys did not match any longer. I fixed it in this test, but this is likely an issue beyond this test whenever `save_on_train_epoch_end` is set to `None` since we restore before this would be evaluated internally.
cc @awaelchli

Reply:
Thanks for investigating guys and sorry, this was caused by a change in master recently (no release affected afaik). Let's take care of this in https://github.com/PyTorchLightning/pytorch-lightning/pull/11108

Reply:
Resolved, fixed in master and updated this branch. Thanks

Reply:
This conversation got a bit sidetracked, but I agree with @rohitgr7 in that the added test overlaps a lot with the modified test

Reply:
So, as the remainder of the discussion here, are we going with a unit test to replace this current test here?
Pseudo code:

```py
cb = ModelCheckpoint(...)
x = {}
cb.on_save_checkpoint(x)

restored_cb = ModelCheckpoint(...)
restored_cb.on_load_checkpoint(x)

for k in ("best_k_models", "kth_best_model_path", "kth_value", "last_model_path"):
        assert cb[k] == getattr(restored_cb, k)
```

No expensive training or saving and loading to disk

Reply:
Thoughs?

Reply:
yes, better remove the old test since it's just checking for ModelCheckpoint states and nothing else.

Reply:
Yes, looks better to me.

Reply:
I replaced the test with a unit-test. Only question I stil have is why not all states returned by `on_save_checkpoint` are restored during `on_load_checkpoint` ([specifcially `current_score`, `dirpath` and `mode`](https://github.com/ORippler/pytorch-lightning/blob/fix/4911/pytorch_lightning/callbacks/model_checkpoint.py#L345-L346)). What I get is that `mode` is used for the state_key generation and therefore equivalence is already checked for this, but what about `current_score' and `dirpath` ?

Reply:
There may be some confusion here. Don't get mislead by state_key, it does not matter.

The current_score does not get restored because you didn't implement that yet. You save it but you don't load it, so you need to fix it.

Reply:
@awaelchli Can you give an update whether `current_score` should be restored ? Currently, I leave it as is and just comment in the test that it's not restored. Might also do a different PR for this, since it's not directly related to #4911

Reply:
As pointed out by @rohitgr7 https://github.com/PyTorchLightning/pytorch-lightning/pull/10995/files#r771482266 it is not required to restore the score.